independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java

```diff
@@ -72,35 +74,52 @@ public ObserverGenerator(AnnotationLiteralProcessor annotationLiterals, Predicat
      * @return a collection of resources
      */
     Collection<Resource> generate(ObserverInfo observer, ReflectionRegistration reflectionRegistration) {
-
-        ClassInfo declaringClass = observer.getObserverMethod().declaringClass();
-        String declaringClassBase;
-        if (declaringClass.enclosingClass() != null) {
-            declaringClassBase = DotNames.simpleName(declaringClass.enclosingClass()) + "_"
-                    + DotNames.simpleName(declaringClass);
+        // The name of the generated class differs:
+        // "org.acme.Foo_Observer_fooMethod_hash" for normal observer where hash represents the signature of the observer method
+        // "org.acme.Registrar_Observer_Synthetic_hash" for synthetic observer where hash represents the basic attrs of the observer
+        String classBase;
+        if (observer.isSynthetic()) {
+            classBase = DotNames.simpleName(observer.getBeanClass());
         } else {
-            declaringClassBase = DotNames.simpleName(declaringClass);
+            ClassInfo declaringClass = observer.getObserverMethod().declaringClass();
+            if (declaringClass.enclosingClass() != null) {
+                classBase = DotNames.simpleName(declaringClass.enclosingClass()) + UNDERSCORE
+                        + DotNames.simpleName(declaringClass);
+            } else {
+                classBase = DotNames.simpleName(declaringClass);
+            }
         }
 
         StringBuilder sigBuilder = new StringBuilder();
-        sigBuilder.append(observer.getObserverMethod().name())
-                .append("_")
-                .append(observer.getObserverMethod().returnType().name().toString());
-        for (org.jboss.jandex.Type paramType : observer.getObserverMethod().parameters()) {
-            sigBuilder.append(paramType.name().toString());
+        if (observer.isSynthetic()) {
+            // This is not unique but best effort
```

Comment:
Hmm if we want to have this unique, we could append object hash of the `ObserverInfo` object? Something like that instead of listing the params. Depends on whether we want to have it human-readable or unique :)

Reply:
Ah, sorry. It should be unique and it should not contain any random element so that it only changes if you change the source. I'm not sure it's 100% needed but use a similar strategy for bean classes where it's definitely needed.

Reply:
Oh, I see what you mean.
I recall we had one issue where synthetic bean re-generated in between tests was already loaded in the CL in its own version. So I suppose this is going to be the same, right?
It was a rare case but I don't think we anyhow solved that (e.g. made it unique on a per-build basis).

Reply:
BTW it might be worth adding something along those lines to the javadoc (of both, observer and bean).